liquid-staking/src/liquidity/add_liquidity.rs

```diff
@@ -0,0 +1,96 @@
+multiversx_sc::imports!();
+
+use crate::{
+    basics,
+    basics::constants::{MINIMUM_LIQUIDITY, MIN_EGLD_TO_DELEGATE, MIN_GAS_FOR_CALLBACK},
+    basics::errors::{
+        ERROR_BAD_PAYMENT_AMOUNT, ERROR_DELEGATION_CONTRACT_NOT_INITIALIZED, ERROR_NOT_ACTIVE,
+    },
+    config, delegation, delegation_proxy, liquidity_pool, StorageCache,
+};
+
+#[multiversx_sc::module]
+pub trait AddLiquidityModule:
+    config::ConfigModule
+    + multiversx_sc_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+    + delegation::DelegationModule
+    + liquidity_pool::LiquidityPoolModule
+    + basics::events::EventsModule
+{
+    #[payable("EGLD")]
+    #[endpoint(addLiquidity)]
+    fn add_liquidity(&self) {
+        self.blockchain().check_caller_is_user_account();
+        let storage_cache = StorageCache::new(self);
+        let caller = self.blockchain().get_caller();
+
+        let payment = self.call_value().egld_value().clone_value();
+        require!(
+            self.is_state_active(storage_cache.contract_state),
+            ERROR_NOT_ACTIVE
+        );
+        if storage_cache.ls_token_supply == 0 {
+            require!(
+                caller == self.blockchain().get_owner_address(),
+                ERROR_DELEGATION_CONTRACT_NOT_INITIALIZED
+            );
+        }
+        require!(payment > MIN_EGLD_TO_DELEGATE, ERROR_BAD_PAYMENT_AMOUNT);
+
+        let delegation_contract = self.get_delegation_contract_for_delegate(&payment);
+
+        let gas_for_async_call = self.get_gas_for_async_call();
+        self.tx()
+            .to(delegation_contract.clone())
+            .typed(delegation_proxy::DelegationMockProxy)
+            .delegate()
+            .egld(payment.clone())
+            .gas(gas_for_async_call)
+            .callback(AddLiquidityModule::callbacks(self).add_liquidity_callback(
+                caller,
+                delegation_contract,
+                payment,
+            ))
+            .gas_for_callback(MIN_GAS_FOR_CALLBACK)
+            .register_promise();
+    }
+
+    #[promises_callback]
+    fn add_liquidity_callback(
+        &self,
+        caller: ManagedAddress,
+        delegation_contract: ManagedAddress,
+        staked_tokens: BigUint,
+        #[call_result] result: ManagedAsyncCallResult<()>,
+    ) {
+        match result {
+            ManagedAsyncCallResult::Ok(()) => {
+                let mut storage_cache = StorageCache::new(self);
+                self.delegation_contract_data(&delegation_contract)
+                    .update(|contract_data| {
+                        contract_data.total_staked_from_ls_contract += &staked_tokens;
+                    });
+
+                let mut ls_token_amount_before_add = BigUint::zero();
```

Comment:
Declare variable closer to where you use it.
```suggestion
                self.delegation_contract_data(&delegation_contract)
                    .update(|contract_data| {
                        contract_data.total_staked_from_ls_contract += &staked_tokens;
                    });

                let mut storage_cache = StorageCache::new(self);
                let mut ls_token_amount_before_add = BigUint::zero();
```